### PR TITLE
Add isoltest option to ignore OS

### LINKED ISSUE #11870: [soltest] Directive for skipping SMTChecker tests on specific platforms
This came up in https://github.com/ethereum/solidity/pull/11861#discussion_r699457642.

## Abstract
If you search the repo for `Disabled because of Spacer nondeterminism`, you'll see that we have quite a few SMTChecker tests where assertions had to be disabled due to results not being deterministic between platforms. The workaround is to comment-out these assertions, which disables them on all platforms and weakens our test suite.

In most cases the results are different just on macOS and I think that it would be better to have a mechanism for skipping the whole test on the specific platforms where we're aware of the non-determinism.

## Specification
Add a soltest directive called `skipPlatforms` that accepts a list of values. Values can be: `macos`, `windows`, `linux`, `emscripten` (I don't think we need more granularity than that). When a test is executed, it should be skipped if the directive contains the name of the current platform.

Example usage:
```solidity
contract C {}
// ====
// SMTEngine: all
// skipPlatforms: macos, windows, linux
// ----
// Warning 6328: (1-5): CHC: Assertion violation might happen here.
// Warning 4661: (1-5): BMC: Assertion violation happens here.
```

### PR DESCRIPTION
**Pull request by leonardoalt**
Fix https://github.com/ethereum/solidity/issues/11870

### CODE REVIEW

File: test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol
Comment (hrkrshnn) [inline test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol:14]: Is this okay?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol:14]: Yea those are just z3 errors when generating counterexamples. What's important is that all OS agree.

**Review by hrkrshnn — DISMISSED** [review]

**Comment by hrkrshnn** [comment]
But maybe wait for the release to end before merging?

**Comment by leonardoalt** [comment]
Yep! Didn't mean for it to go in before

**Comment by hrkrshnn** [comment]
https://github.com/ethereum/solidity/pull/12038 would conflict with this.

File: test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol
Comment (christianparpart) [inline test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol:6]: While I do not disagree with this change. I wonder. Shouldn't we have a PR for one subject and another for something else? (bad wording maybe. jsut not having too much different things in one PR / commit).

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol:6]: The code change is there exactly to allow re-enabling such tests.

File: test/libsolidity/SMTCheckerTest.cpp
Comment (cameel) [inline test/libsolidity/SMTCheckerTest.cpp]: It's a bit confusing to call the whole string `os`.
```suggestion
	string const& ignoreOSSetting = m_reader.stringSetting("SMTIgnoreOS", "none");
	for (string const& os: ignoreOSSetting | ranges::views::split(',') | ranges::to<vector<string>>())
```

File: test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol
Comment (cameel) [reply test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol:6]: Yeah, the whole point was that these had to be disabled becuse they were failing on macOS so uncommenting them shows that the new setting works.

**Review by cameel — DISMISSED** [review]
I'd put it in `SyntaxTest` rather than make it SMT specific but, well, only SMTChecker needs it now so I guess we can extend it later if we need it elsewhere.

File: test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/external_calls/call_with_value_1.sol:6]: I'll still enable more tests in a different PR, but wanted to enable at least one back in this one to prove the point and not bloat it.

**Comment by leonardoalt** [comment]
Fixed code

**Comment by leonardoalt** [comment]
@cameel can you reapprove?

**Review by cameel — APPROVED** [review]

**Pull request merged**
